panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,714 @@
+  nm_setting_user_set_data (setting_user, NM_SETTING_TARIFF, tariff_string, &error);
+
+  if (error)
+    {
+      g_warning ("Error storing "NM_SETTING_TARIFF": %s", error->message);
+      return;
+    }
+
+  nm_remote_connection_commit_changes_async (NM_REMOTE_CONNECTION (connection),
+                                             TRUE, /* save to disk */
+                                             NULL,
+                                             on_network_changes_committed_cb,
+                                             NULL);
+
+  /* Make sure the application does not exit before the callback is called */
+  g_application_hold (g_application_get_default ());
+}
+
+static void
+get_active_connection_and_device (CcUpdatesPanel  *self,
+                                  NMDevice       **out_device,
+                                  NMConnection   **out_connection,
+                                  NMAccessPoint  **out_ap)
+{
+  NMActiveConnection *active_connection = NULL;
+  NMConnection *connection = NULL;
+  const GPtrArray *active_devices = NULL;
+  NMAccessPoint *ap = NULL;
+  NMDevice *active_device = NULL;
+
+  active_connection = nm_client_get_primary_connection (self->nm_client);
+
+  /* If no primary connection is already present, try and use the connecting one */
+  if (!active_connection)
+    active_connection = nm_client_get_activating_connection (self->nm_client);
+
+  if (active_connection)
+    {
+      connection = NM_CONNECTION (nm_active_connection_get_connection (active_connection));
+      active_devices = nm_active_connection_get_devices (active_connection);
+
+      if (active_devices && active_devices->len > 0)
+        {
+          /* This array is guaranteed to have only one element */
+          active_device = g_ptr_array_index (active_devices, 0);
+
+          if (NM_IS_DEVICE_WIFI (active_device))
+            ap = nm_device_wifi_get_active_access_point (NM_DEVICE_WIFI (active_device));
+        }
+    }
+
+  if (out_device)
+    *out_device = active_device;
+
+  if (out_connection)
+    *out_connection = connection;
+
+  if (out_ap)
+    *out_ap = ap;
+}
+
+static void
+ensure_setting_user (CcUpdatesPanel *self,
+                     NMConnection   *connection)
+{
+  NMSettingUser *setting_user;
+
+  if (!connection)
+    return;
+
+  setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+
+  if (!setting_user)
+    {
+      g_autoptr(GError) error = NULL;
+      NMSettingConnection *setting;
+      NMMetered metered;
+      gboolean enabled;
+
+      g_debug ("Creating a new custom config file for the current connection…");
+
+      /* Add a new NMSettingUser to this connection */
+      setting_user = NM_SETTING_USER (nm_setting_user_new ());
+      nm_connection_add_setting (connection, NM_SETTING (setting_user));
+
+      /* The default value depends on the metered state of the connection */
+      setting = nm_connection_get_setting_connection (connection);
```

Comment:
I think `nm_connection_get_setting_connection()` can also potentially return `NULL`.

Reply:
I’ve left a follow-up comment on the `ensure_setting_connection()` stuff.